### PR TITLE
Fixed #32076 -- Added async methods to BaseCache.

### PR DESCRIPTION
**Pull request by Andrew-Chen-Wang**
[Ticket-32076](https://code.djangoproject.com/ticket/32076)

Per DEP 9, the async methods of the cache default to running the synchronous version.

### CODE REVIEW

**Review by felixxm — COMMENTED** [review]
@Andrew-Chen-Wang Thanks for this patch :+1: 

We've [discussed](https://forum.djangoproject.com/t/naming-of-async-variants/7216) and reached a consensus about naming async variants of methods. Please use the `a` prefix instead of `_async` suffix, i.e.: `aadd()`, `aget()`, `aset()`, `atouch()`, `adelete()`, `aget_many()`, `aget_or_set()`, `ahas_key()`, `aincr()`, `adecr()`, `aset_many()`, `adelete_many()`, `aclear()`, `aincr_version()`, `adecr_version()`, and `aclose()`.

We should also document this changes, at least release notes and a small section (e.g. _"Asynchronous support"_) in [_"Django’s cache framework"_](https://docs.djangoproject.com/en/3.1/topics/cache/#django-s-cache-framework) are needed.

**Comment by Andrew-Chen-Wang** [comment]
@felixxm I like the prefix! I've updated the PR with the prefix, docs, and release note.

File: django/core/cache/backends/dummy.py [hunk 1]
```diff
@@ -12,28 +12,49 @@ def add(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
...
         return False
 
+    async def ahas_key(self, *args, **kwargs):
+        return self.has_key(*args, **kwargs)
+
     def clear(self):
         pass
+
+    async def aclear(self):
+        pass
```
Comment (felixxm) [inline django/core/cache/backends/dummy.py:60]: All these transparent methods in the `DummyCache` wouldn't be necessary if we change `BaseCache` implementation from:
```python
aadd = sync_to_async(add, thread_sensitive=True)
```
to
```python
async def aadd(self, *args, **kwargs):
    return await sync_to_async(self.add, thread_sensitive=True)(*args, **kwargs)
```
@andrewgodwin What do you think?

File: tests/cache/tests.py [hunk 2]
```diff
@@ -88,28 +89,55 @@ def test_simple(self):
         cache.set("key", "value")
         self.assertIsNone(cache.get("key"))
 
+    async def test_simple_async(self):
+        "Dummy cache backend ignores cache set calls"
+        await cache.aset("key", "value")
+        self.assertIsNone(await cache.aget("key"))
```
Comment (felixxm) [inline tests/cache/tests.py:95]: I think we should move all async tests to a new file, e.g. `tests_async.py` with `AsyncDummyCacheTests`.

**Review by felixxm — COMMENTED** [review]
@Andrew-Chen-Wang Thanks for updates :+1:

**Comment by felixxm** [comment]
@Andrew-Chen-Wang  Do you have time to keep working on this?

**Comment by Andrew-Chen-Wang** [comment]
@felixxm yes, I should have time over the US's Labor Day weekend (this weekend). Was just waiting for a response pertaining to your comment here https://github.com/django/django/pull/13547#discussion_r637837628 from @andrewgodwin .

File: django/core/cache/backends/dummy.py (hunk 1 above)
Comment (andrewgodwin) [reply django/core/cache/backends/dummy.py:60]: While it's less clean on the base side, I think it's a sensible change - will mean that most cache implementations "just work" due to the lazier binding style.

**Comment by andrewgodwin** [comment]
I am liking seeing the `a` prefix in action, by the way - I think this solidifies it as a style choice. Thanks for working on this!

**Comment by Andrew-Chen-Wang** [comment]
@felixxm rebased. Thanks for the quick responses! Note: I also moved the release docs for the cache to 4.0.

File: docs/releases/3.2.txt [hunk 3]
```diff
@@ -286,6 +286,11 @@ Minor features
 * The new ``item_comments`` hook allows specifying a comments URL per feed
   item.
 
+Cache
+~~~~~
+
+* ...
+
```
Comment (felixxm) [inline docs/releases/3.2.txt:293]: Revert this change.

File: django/core/cache/backends/dummy.py [hunk 4]
```diff
@@ -12,29 +12,50 @@ def add(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
         self.validate_key(key)
         return True
 
+    async def aadd(self, *args, **kwargs):
```
Comment (felixxm) [inline django/core/cache/backends/dummy.py:15]: Dummy backend no longer needs new methods after changing `BaseCache` implementation as proposed in the [comment](https://github.com/django/django/pull/13547/files#r637837628). Please revert changes in the `DummyCache`.

File: django/core/cache/backends/base.py [hunk 5]
```diff
@@ -115,34 +117,54 @@ def add(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
         """
         raise NotImplementedError('subclasses of BaseCache must provide an add() method')
 
+    async def aadd(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
+        """See add()."""
```
Comment (felixxm) [inline django/core/cache/backends/base.py:121]: I don't see much value in `See method().` docstrings when it's only `sync_to_async()` call. Please remove it from `aadd`, `aget`, `aset`, `atouch`, `adelete`, `ahas_key`, `adecr`, `aclear`, and `aclose`.

**Review by felixxm — COMMENTED** [review]
@Andrew-Chen-Wang Thanks for updates :+1:

File: django/core/cache/backends/dummy.py (hunk 4 above)
Comment (Andrew-Chen-Wang) [reply django/core/cache/backends/dummy.py:15]: Seems inefficient though, no?

**Review by felixxm — COMMENTED** [review]
@Andrew-Chen-Wang Thanks again :+1: Welcome aboard :boat: 

I rebased, squashed commits, and pushed small to tests and docs.

File: docs/releases/4.0.txt [hunk 6]
```diff
@@ -187,7 +187,11 @@ Minor features
 Cache
 ~~~~~
 
-* ...
+* The new async API for ``django.core.cache.backends.base.BaseCache`` begins
+  process of making cache backends async-compatible. All new methods have the
+  ``a`` prefixed names, e.g. ``aadd()``, ``aget()``, ``aset()``,
+  ``aget_or_set()``, or ``adelete_many()``. The ``a`` prefix will be used in
+  the future for async variants of any methods.
```
Comment (carltongibson) [inline docs/releases/4.0.txt:194]: `begins the process` but, maybe... 

```suggestion
* The new async API for ``django.core.cache.backends.base.BaseCache`` begins 
  the process of making cache backends async-compatible. The new async methods
  all have ``a`` prefixed names, e.g. ``aadd()``, ``aget()``, ``aset()``,
  ``aget_or_set()``, or ``adelete_many()``. 
  
  Going forward, the ``a`` prefix will be used for async variants of
  methods generally.
```

File: docs/topics/cache.txt [hunk 7]
```diff
@@ -1123,6 +1135,25 @@ instance, to do this for the ``locmem`` backend, put this code in a module::
...
+
+.. versionadded:: 4.0
+
+Django has developing support for asynchronous cache backends, but does not
+yet support asynchronous caching. It will be coming in a future release.
+
+``django.core.cache.backends.base.BaseCache`` has async variants of
+:ref:`all base methods <cache-basic-interface>`. To prevent confusion between
+synchronous and asynchronous naming, all asynchronous methods are prefixed
+with ``a``. By default, the arguments for both variants are the same::
```
Comment (carltongibson) [inline docs/topics/cache.txt:1151]: Maybe: 

```suggestion
``django.core.cache.backends.base.BaseCache`` has async variants of :ref:`all
base methods <cache-basic-interface>`. By convention, the asynchronous versions
of all methods are prefixed with ``a``. By default, the arguments for both
variants are the same::
```

Is it really to _prevent confusion_ …? — or is it about _awaitables_ vs regular functions, and then naming conventions on top of that? 🤔 (Not sure)

(In any case, is the wrapping quite right?)

**Review by carltongibson — APPROVED** [review]
Just a couple of comment on the docs changes. 

👍

File: docs/releases/4.0.txt (hunk 6 above)
Comment (Andrew-Chen-Wang) [inline docs/releases/4.0.txt:194]: I'm confused are we still keeping the docs for 4.0? What happened to the revert on the 3.2 docs?

Comment (Andrew-Chen-Wang) [reply docs/releases/4.0.txt:194]: @felixxm I'm 👍 with this.

File: docs/topics/cache.txt (hunk 7 above)
Comment (Andrew-Chen-Wang) [reply docs/topics/cache.txt:1151]: I'm 👍 with this. I didn't realize you could wrap the `:ref:` on multiple lines

File: docs/releases/4.0.txt (hunk 6 above)
Comment (felixxm) [reply docs/releases/4.0.txt:194]: It's a new features so it will included in the Django 4.0. Previously I requested to revert adding and empty _"Cache"_ section to the Django 3.2. release notes.

**Comment by felixxm** [comment]
@carltongibson Thanks for the review :+1: Docs updated.
